From 32373d6b807ad5637788ce47e1174da6d196f094 Mon Sep 17 00:00:00 2001
From: tatataeki <shengzeyu19_98@163.com>
Date: Wed, 29 Sep 2021 18:49:51 +0800
Subject: [PATCH 1/4] signal monitor initialization add some basic statements
 essential for the communication and add the function to initiate the data
 structure of netlink for communication

Signed-off-by: tatataeki <shengzeyu19_98@163.com>
---
 kernel/signal.c | 40 ++++++++++++++++++++++++++++++++++++++++
 1 file changed, 40 insertions(+)

diff --git a/kernel/signal.c b/kernel/signal.c
index ef8f2a28d37c..e0175bf72463 100644
--- a/kernel/signal.c
+++ b/kernel/signal.c
@@ -56,6 +56,8 @@
 #include <asm/siginfo.h>
 #include <asm/cacheflush.h>
 
+#include <net/sock.h>
+#include <net/netlink.h>
 /*
  * SLAB caches for signal bits.
  */
@@ -4637,3 +4639,41 @@ void kdb_send_sig(struct task_struct *t, int sig)
 		kdb_printf("Signal %d is sent to process %d.\n", sig, t->pid);
 }
 #endif	/* CONFIG_KGDB_KDB */
+
+#define NETLINK_SIG_MONITOR 25
+
+struct sock *sig_monitor_sk = NULL;
+
+extern struct net init_net;
+
+void recv_callback(struct sk_buff *skb);
+
+struct netlink_kernel_cfg sig_cfg = {
+	.input = recv_callback,
+};
+
+int netlink_sig_init(void)
+{
+	sig_monitor_sk = (struct sock*)netlink_kernel_create(&init_net, NETLINK_SIG_MONITOR, &sig_cfg);
+    	if(sig_monitor_sk == NULL) {
+        	return 1;
+    	}
+	return 0;
+}
+
+/*
+ * prepare_for_nt_send - Prepare for the netlink struct to send signal
+ * info to user program. The function below check the info in the signal
+ * list when handling them, and send the filtered signal info.
+ */
+void prepare_for_nt_send(void) {
+	if(!sig_monitor_sk) {
+		int ret = netlink_sig_init();
+		if(ret == 1) {
+			printk(KERN_ERR "SIGNAL MONITOR: create netlink socket error.\n");
+		} else {
+			printk("SIGNAL MONITOR: create netlink socket ok.\n");
+		}
+	}
+}
+
-- 
2.27.0

